File: server/bedrock/src/NovaEdge/Engine/Matchmaking.php
<?php
// ═══════════════════════════════════════════════════════════════════
//  Nova Edge — Engine/Matchmaking.php (پورت دقیق shared/engine/matchmaking.js)
//  ⚠️ ترتیب مصرف RNG باید دقیقاً مثل JS باشد تا نام/رتبهٔ بات‌ها یکسان دربیاید.
// ═══════════════════════════════════════════════════════════════════
declare(strict_types=1);

namespace NovaEdge\Engine;

final class Matchmaking
{
    public const BOT_NAMES = [
        'Arash_77', 'NimaX', 'SinaPvP', 'KianCraft', 'ParsaGod', 'MiladYT', 'AmirTNT', 'RezaBridge',
        'Saman_98', 'Tohid', 'BardiaPro', 'ErfanZ', 'MahdiSword', 'AlirezaMC', 'HosseinBed', 'JavadRun',
        'NaderSky', 'YasinPvP', 'Mobin_1', 'Soroush', 'FarhadGG', 'KamyarX', 'PeymanUHC', 'Bahram',
        'AriaParkour', 'ShayanKit', 'OmidDuel', 'VahidSpleef', 'MortezaAim', 'SaeedTNT', 'DanialMurder',
        'ImanBuilder', 'AshkanFrost', 'MehdiVamp', 'NavidRush', 'ZahraMC', 'SaraPvP', 'NargesBuild',
        'LeilaSky', 'MaryamGG', 'ElnazRun', 'FatemehKit', 'HaniehDuel', 'RoxanaX', 'TaraMC', 'YasnaPro',
        'Alex_Steve', 'Notch_Fan', 'EnderPro', 'CreeperHug', 'RedstoneRat', 'PistonPete', 'SlimeKing',
        'BlazeBorn', 'IronGolemX', 'WitherWatch', 'GhastGoal', 'ShulkerSam', 'PiglinPal', 'WardenWake',
    ];

    public static function botName(int $i, ?Rng $rng): string
    {
        $list = self::BOT_NAMES;
        $base = $list[$i % count($list)];
        $suffix = '';
        if ($rng) {
            if ($rng->chance(0.35)) {
                $suffix = '_' . $rng->int(10, 99);
            }
        }
        return $base . $suffix;
    }

    /** @param array $o {index, tier, matchRating, mode, rng, platform} */
    public static function makeBot(array $o): array
    {
        $rng = $o['rng'] ?? null;
        $name = self::botName((int) ($o['index'] ?? 0), $rng);
        $jitter = $rng ? $rng->int(-90, 90) : 0;
        $rating = RngCore::clamp(Bits::jsRound((float) ($o['matchRating'] ?? 1000) + $jitter), 100, 4000);
        $mode = $o['mode'] ?? null;
        return [
            'id' => 'bot_' . preg_replace('/[^a-z0-9]/', '', strtolower($name)) . '_' . (int) ($o['index'] ?? 0),
            'name' => $name,
            'is_bot' => true,
            'tier' => $o['tier'] ?? null,
            'rating' => $rating,
            'level' => $rng ? $rng->int(5, 90) : 20,
            'rank_id' => 'free',
            'platform' => $o['platform'] ?? ($rng && $rng->chance(0.35) ? 'bedrock' : 'java'),
            'games' => $rng ? $rng->int(20, 900) : 100,
            'seed' => $rng ? $rng->int(1, 2147483647) : (int) ($o['index'] ?? 0) + 1,
            'mode' => is_array($mode) ? ($mode['id'] ?? null) : $mode,
        ];
    }

    private static function sumRating(array $players): float
    {
        $s = 0.0;
        foreach ($players as $p) {
            $s += (float) ($p['rating'] ?? 0) ?: 1000;
        }
        return $s;
    }

    /** @param array $teams */
    private static function pickPlatform(array $weakest, array $teams, mixed $crossplay, Rng $rng): string
    {
        if (!$crossplay) {
            return $rng->chance(0.5) ? 'java' : 'bedrock';
        }
        $java = 0;
        $bedrock = 0;
        foreach ($teams as $t) {
            foreach ($t['players'] as $p) {
                if (($p['platform'] ?? '') === 'bedrock') {
                    $bedrock++;
                } else {
                    $java++;
                }
            }
        }
        return $java <= $bedrock ? 'java' : 'bedrock';
    }

    /**
     * بالانس تیم‌ها: درفت مارپیچی + بات‌ها به ضعیف‌ترین تیم
     * @param array $humans @param array $opts {mode, seed, tier, crossplay}
     */
    public static function balanceTeams(array $humans, int $botsNeeded, int $teamCount, int $teamSize, array $opts = []): array
    {
        $rng = new Rng((int) ($opts['seed'] ?? 1));
        $mode = $opts['mode'] ?? [];
        $colors = ['red', 'blue', 'green', 'yellow', 'aqua', 'pink', 'gray', 'orange', 'white', 'purple', 'brown', 'lime'];
        $pool = $humans;
        usort($pool, fn ($a, $b) => (((float) ($b['rating'] ?? 0) ?: 1000)) <=> (((float) ($a['rating'] ?? 0) ?: 1000)));
        $avgRating = $pool ? array_sum(array_map(fn ($p) => (float) ($p['rating'] ?? 0) ?: 1000, $pool)) / count($pool) : 1000;

        $teams = [];
        $cat = $mode['category'] ?? '';
        $nTeams = in_array($cat, ['ffa', 'race', 'creative'], true)
            ? max(1, count($pool) + $botsNeeded)
            : max(1, min($teamCount ?: 2, max(2, (int) ceil((count($pool) + $botsNeeded) / max(1, $teamSize ?: 1)))));

        for ($i = 0; $i < $nTeams; $i++) {
            $teams[] = ['index' => $i, 'color' => $colors[$i % count($colors)], 'players' => [], 'avg_rating' => 0, 'platforms' => ['java' => 0, 'bedrock' => 0]];
        }

        $dir = 1;
        $cur = 0;
        foreach ($pool as $p) {
            $teams[$cur]['players'][] = $p;
            $cur += $dir;
            if ($cur >= count($teams)) {
                $cur = count($teams) - 1;
                $dir = -1;
            } elseif ($cur < 0) {
                $cur = 0;
                $dir = 1;
            }
        }

        $matchRating = $avgRating;
        for ($i = 0; $i < $botsNeeded; $i++) {
            // ضعیف‌ترین تیم (اولین حداقل در صورت تساوی — مثل sort پایدار JS) بدون برهم‌زدن ترتیب تیم‌ها
            $wi = 0;
            $wv = self::sumRating($teams[0]['players']);
            foreach ($teams as $ti => $t) {
                $sv = self::sumRating($t['players']);
                if ($sv < $wv) {
                    $wv = $sv;
                    $wi = $ti;
                }
            }
            $platform = self::pickPlatform($teams[$wi], $teams, $opts['crossplay'] ?? null, $rng);
            $teams[$wi]['players'][] = self::makeBot(['index' => $i, 'tier' => $opts['tier'] ?? 'T0', 'matchRating' => $matchRating, 'mode' => $mode, 'rng' => $rng, 'platform' => $platform]);
        }

        foreach ($teams as &$t) {
            $t['avg_rating'] = (int) Bits::jsRound(self::sumRating($t['players']) / max(1, count($t['players'])));
            foreach ($t['players'] as $p) {
                $t['platforms'][($p['platform'] ?? '') === 'bedrock' ? 'bedrock' : 'java']++;
            }
        }
        return $teams;
    }
}

File: server/java/src/main/java/ir/novaedge/core/Matchmaking.java
// ═══════════════════════════════════════════════════════════════════
//  Nova Edge — core/Matchmaking.java (پورت دقیق shared/engine/matchmaking.js)
// ═══════════════════════════════════════════════════════════════════
package ir.novaedge.core;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;

public final class Matchmaking {

    public static final String[] BOT_NAMES = {
            "Arash_77", "NimaX", "SinaPvP", "KianCraft", "ParsaGod", "MiladYT", "AmirTNT", "RezaBridge",
            "Saman_98", "Tohid", "BardiaPro", "ErfanZ", "MahdiSword", "AlirezaMC", "HosseinBed", "JavadRun",
            "NaderSky", "YasinPvP", "Mobin_1", "Soroush", "FarhadGG", "KamyarX", "PeymanUHC", "Bahram",
            "AriaParkour", "ShayanKit", "OmidDuel", "VahidSpleef", "MortezaAim", "SaeedTNT", "DanialMurder",
            "ImanBuilder", "AshkanFrost", "MehdiVamp", "NavidRush", "ZahraMC", "SaraPvP", "NargesBuild",
            "LeilaSky", "MaryamGG", "ElnazRun", "FatemehKit", "HaniehDuel", "RoxanaX", "TaraMC", "YasnaPro",
            "Alex_Steve", "Notch_Fan", "EnderPro", "CreeperHug", "RedstoneRat", "PistonPete", "SlimeKing",
            "BlazeBorn", "IronGolemX", "WitherWatch", "GhastGoal", "ShulkerSam", "PiglinPal", "WardenWake",
    };

    private Matchmaking() {
    }

    public static String botName(int i, Rng.Generator rng) {
        String base = BOT_NAMES[Math.floorMod(i, BOT_NAMES.length)];
        String suffix = "";
        if (rng != null && rng.chance(0.35)) {
            suffix = "_" + rng.int_(10, 99);
        }
        return base + suffix;
    }

    public static Map<String, Object> makeBot(int index, String tier, double matchRating, Map<String, Object> mode,
            Rng.Generator rng, String platform) {
        String name = botName(index, rng);
        int jitter = rng != null ? rng.int_(-90, 90) : 0;
        double rating = Rng.clamp(Rng.jsRound(matchRating + jitter), 100, 4000);
        Map<String, Object> out = Json.obj();
        Json.put(out, "id", "bot_" + name.toLowerCase().replaceAll("[^a-z0-9]", "") + "_" + index);
        Json.put(out, "name", name);
        Json.put(out, "is_bot", true);
        Json.put(out, "tier", tier);
        Json.put(out, "rating", rating);
        Json.put(out, "level", rng != null ? rng.int_(5, 90) : 20);
        Json.put(out, "rank_id", "free");
        Json.put(out, "platform", platform != null ? platform : (rng != null && rng.chance(0.35) ? "bedrock" : "java"));
        Json.put(out, "games", rng != null ? rng.int_(20, 900) : 100);
        Json.put(out, "seed", rng != null ? rng.int_(1, 2147483647) : index + 1);
        Json.put(out, "mode", mode == null ? null : mode.get("id"));
        return out;
    }

    private static double sumRating(List<Map<String, Object>> players) {
        double s = 0;
        for (Map<String, Object> p : players) {
            double r = Json.num(p, "rating", 0);
            s += r != 0 ? r : 1000;
        }
        return s;
    }

    private static String pickPlatform(Map<String, Object> weakest, List<Map<String, Object>> teams, Object crossplay, Rng.Generator rng) {
        if (!Json.truthy(crossplay)) {
            return rng.chance(0.5) ? "java" : "bedrock";
        }
        int java = 0;
        int bedrock = 0;
        for (Map<String, Object> t : teams) {
            for (Map<String, Object> p : Json.mapList(t.get("players"))) {
                if ("bedrock".equals(Json.str(p, "platform", ""))) {
                    bedrock++;
                } else {
                    java++;
                }
            }
        }
        return java <= bedrock ? "java" : "bedrock";
    }

    @SuppressWarnings("unchecked")
    public static List<Map<String, Object>> balanceTeams(List<Map<String, Object>> humans, int botsNeeded,
            int teamCount, int teamSize, Map<String, Object> opts) {
        Rng.Generator rng = new Rng.Generator((int) Json.num(opts, "seed", 1));
        Map<String, Object> mode = Json.map(opts, "mode");
        String[] colors = {"red", "blue", "green", "yellow", "aqua", "pink", "gray", "orange", "white", "purple", "brown", "lime"};
        List<Map<String, Object>> pool = new ArrayList<>(humans);
        pool.sort(Comparator.comparingDouble((Map<String, Object> p) -> {
            double r = Json.num(p, "rating", 0);
            return r != 0 ? r : 1000;
        }).reversed());
        double avg = 1000;
        if (!pool.isEmpty()) {
            double s = 0;
            for (Map<String, Object> p : pool) {
                double r = Json.num(p, "rating", 0);
                s += r != 0 ? r : 1000;
            }
            avg = s / pool.size();
        }

        List<Map<String, Object>> teams = new ArrayList<>();
        String cat = mode == null ? "" : Json.str(mode, "category", "");
        int nTeams = cat.equals("ffa") || cat.equals("race") || cat.equals("creative")
                ? Math.max(1, pool.size() + botsNeeded)
                : Math.max(1, Math.min(teamCount != 0 ? teamCount : 2,
                        Math.max(2, (int) Math.ceil((pool.size() + botsNeeded) / (double) Math.max(1, teamSize)))));
        for (int i = 0; i < nTeams; i++) {
            Map<String, Object> t = Json.obj();
            Json.put(t, "index", i);
            Json.put(t, "color", colors[i % colors.length]);
            Json.put(t, "players", new ArrayList<Map<String, Object>>());
            Json.put(t, "avg_rating", 0);
            Map<String, Object> pf = Json.obj();
            pf.put("java", 0);
            pf.put("bedrock", 0);
            Json.put(t, "platforms", pf);
            teams.add(t);
        }

        int dir = 1;
        int cur = 0;
        for (Map<String, Object> p : pool) {
            ((List<Map<String, Object>>) teams.get(cur).get("players")).add(p);
            cur += dir;
            if (cur >= teams.size()) {
                cur = teams.size() - 1;
                dir = -1;
            } else if (cur < 0) {
                cur = 0;
                dir = 1;
            }
        }

        for (int i = 0; i < botsNeeded; i++) {
            int wi = 0;
            double wv = sumRating((List<Map<String, Object>>) teams.get(0).get("players"));
            for (int ti = 0; ti < teams.size(); ti++) {
                double sv = sumRating((List<Map<String, Object>>) teams.get(ti).get("players"));
                if (sv < wv) {
                    wv = sv;
                    wi = ti;
                }
            }
            String platform = pickPlatform(teams.get(wi), teams, opts.get("crossplay"), rng);
            ((List<Map<String, Object>>) teams.get(wi).get("players")).add(
                    makeBot(i, Json.str(opts, "tier", "T0"), avg, mode, rng, platform));
        }

        for (Map<String, Object> t : teams) {
            List<Map<String, Object>> ps = (List<Map<String, Object>>) t.get("players");
            Json.put(t, "avg_rating", (long) Rng.jsRound(sumRating(ps) / Math.max(1, ps.size())));
            Map<String, Object> pf = (Map<String, Object>) t.get("platforms");
            pf.put("java", 0);
            pf.put("bedrock", 0);
            for (Map<String, Object> p : ps) {
                String k = "bedrock".equals(Json.str(p, "platform", "")) ? "bedrock" : "java";
                pf.put(k, ((Number) pf.get(k)).intValue() + 1);
            }
        }
        return teams;
    }
}
